### PR TITLE
Align Hopper FP8 blockwise quant to power-of-2 scales; fix tl.trans

### PR DESCRIPTION
**Pull request by haok1402**
**TLDR**: Brings the Hopper FP8 path onto the same power-of-2 block scaling that NVIDIA TransformerEngine uses by default. In TE, fp32-mantissa scales are a Hopper-only opt-out (`NVTE_FP8_BLOCK_SCALING_FP32_SCALES=1`) and Blackwell is pow2-only. See [FP8 Blockwise Scaling](https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/features/low_precision_training/fp8_blockwise_scaling/fp8_blockwise_scaling.html).

## What changed
- **Fixed a correctness bug:** Triton 3.7.1 `tl.trans` corrupts `fp8` tiles, producing a wrong transposed FP8 weight (used in dgrad). Fix: store to transposed addresses instead of `tl.trans` on fp8.
- **Renamed the functions** `fused_* -> fp8cast_*`: every kernel in this module is fused, so `fused_` said nothing; `fp8cast_` uses DeepGEMM's own "cast" verb.
- **Renamed the scale flag** `SCALING_MODE` -> `NATIVE_E8M0` (it just means native-SM100 vs emulated-SM90 E8M0; both are pow2) and corrected the docstrings.
- **Tests:** moved `tests/test_fp8_quantize_kernels.py` -> `tests/operators/test_deepgemm_quantize.py`; fixed the reference to always use E8M0 (was previously device-gated and wrong on Hopper -> 39 failures).

## Validation
- `pytest tests/operators/test_deepgemm_quantize.py` -> **48 passed** (was 39 failing on H100).
- Minimal standalone repro of the Triton bug (`workspace/triton_fp8_trans_repro.py`, included below): `kernel1` (`tl.trans`) mismatches 4090/16384; `kernel2` (transposed-address store, the fix) passes with 0.
- dsv2-lite FP8 training run to confirm the loss curve is unchanged vs. reference (fix only touches dgrad).

## Files
- `pithtrain/operators/deepgemm_quantize.py`
- `tests/operators/test_deepgemm_quantize.py`

## Reproduction of the Triton bug

```python
"""
On an NVIDIA H100 80GB HBM3 (Python 3.14.3, Triton 3.7.1, torch 2.12.1+cu130),
the following outputs were observed:

kernel1 (tl.trans):      mismatch 4090 / 16384
kernel2 (address store): mismatch 0 / 16384
"""

import torch
import triton
import triton.language as tl

BLOCK = 128


@triton.jit
def kernel1(x_ptr, o_ptr, t_ptr, BLOCK: tl.constexpr = BLOCK):
    r = tl.arange(0, BLOCK)
    off = r[:, None] * BLOCK + r[None, :]
    fp8 = tl.load(x_ptr + off).to(tl.float32).to(tl.float8e4nv)
    tl.store(o_ptr + off, fp8)
    tl.store(t_ptr + off, tl.trans(fp8))


@triton.jit
def kernel2(x_ptr, o_ptr, t_ptr, BLOCK: tl.constexpr = BLOCK):
    r = tl.arange(0, BLOCK)
    off = r[:, None] * BLOCK + r[None, :]
    tsp = r[None, :] * BLOCK + r[:, None]
    fp8 = tl.load(x_ptr + off).to(tl.float32).to(tl.float8e4nv)
    tl.store(o_ptr + off, fp8)
    tl.store(t_ptr + tsp, fp8)


def run(kernel):
    x = torch.randn(BLOCK, BLOCK, device="cuda", dtype=torch.bfloat16)
    o = torch.empty(BLOCK, BLOCK, device="cuda", dtype=torch.float8_e4m3fn)
    t = torch.empty(BLOCK, BLOCK, device="cuda", dtype=torch.float8_e4m3fn)
    kernel[(1,)](x, o, t)
    return (t.float() != o.float().T).sum().item()


print("kernel1 (tl.trans):      mismatch", run(kernel1), "/", BLOCK * BLOCK)
print("kernel2 (address store): mismatch", run(kernel2), "/", BLOCK * BLOCK)
```

### CODE REVIEW

**Comment by haok1402** [comment]
@claude review

File: pithtrain/operators/deepgemm_quantize.py [hunk 1]
```diff
@@ -92,12 +82,12 @@ def _compute_fp8_scale(amax, SCALING_MODE: tl.constexpr):
...
 
 # ---------------------------------------------------------------------------
-# fused_rowwise_colwise: rowwise(x) + colwise(x)
+# fp8cast_rowwise_colwise: rowwise(x) + colwise(x)
 # ---------------------------------------------------------------------------
 
 
 @triton.jit
-def _fused_rowwise_colwise_fp8_kernel(
+def fp8cast_rowwise_colwise_kernel(
```
Comment (MasterJH5574) [inline pithtrain/operators/deepgemm_quantize.py]: keep the underscore since this function is not intended to be public.

Comment (haok1402) [reply pithtrain/operators/deepgemm_quantize.py]: @MasterJH5574 Fixed. Internal fused kernels are now underscored with `_` and we add the `__all__` to unambiguously specify all the exported usage.

**Comment by haok1402** [comment]
> Performance Review

Benchmarked the three transposed-store strategies on an idle H100 (`do_bench`). No measurable gap.

```
         shape |       trans (base) |      address (cur) |   bf16 (coalesced)
     2048x2048 |   11.5us  1460GB/s |   11.4us  1468GB/s |   11.4us  1476GB/s |
     4096x4096 |   30.1us  2232GB/s |   29.7us  2259GB/s |   29.8us  2254GB/s |
     7168x2048 |   27.0us  2172GB/s |   26.8us  2191GB/s |   26.3us  2232GB/s |
     2048x7168 |   27.0us  2171GB/s |   26.9us  2185GB/s |   26.8us  2192GB/s |
     8192x8192 |  101.1us  2654GB/s |  101.2us  2652GB/s |  100.7us  2666GB/s |
```

All three within ~1%. No regression.

Also compared end-to-end training throughput (dsv2-lite, 64 steps, steady-state):

```
              | before (tl.trans) | after (address-store)
tokens/sec    |       54,914      |        55,356
step-time (s) |        38.19      |         37.89
```

+0.81% after. Within noise. No regression at the training level either.

**Comment by haok1402** [comment]
> The committed evidence is the unit suite ("48 passed"), which does validate the transposed FP8 output bit-exactly against the reference — strong for the kernel. But the claimed "dsv2-lite FP8 training run to confirm the loss curve is unchanged" is asserted only; no loss table, log, or wandb link is committed, and the standalone repro referenced as workspace/triton_fp8_trans_repro.py is not in the branch (only inlined in the PR body).

image

Fixed. See https://wandb.ai/pithtrain/hopper-deepgemm/workspace?nw=28pacu7obnf

**Comment by haok1402** [comment]
@claude review

**Review by MasterJH5574 — APPROVED** [review]
